File: 23. Merge k Sorted Lists/solution-heap.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        # assume there are m list with n nodes each on average
        # create a min heap for each node's value, O(mn*log(mn))
        heap = []
        for head in lists:
            while head:
                heapq.heappush(heap, head.val)
                head = head.next
        
        # pop elements from heap and add it to a linked list
        # total elements = mn, heappop = log(mn), Complexity = O(mn*log(mn))
        head = ListNode(None)
        node = head
        while heap:
            node.next = ListNode(heapq.heappop(heap))
            node = node.next
        
        return head.next

File: 23. Merge k Sorted Lists/solution-heap.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = new ListNode(Integer.MIN_VALUE);
        ListNode node = head;
        PriorityQueue<ListNode> heap = new PriorityQueue<>(11, (x, y) -> x.val - y.val);
        for(ListNode nodes: lists) {
            while (nodes != null) {
                node = nodes;
                nodes = nodes.next;
                node.next = null;
                heap.add(node);
            }
        }
        
        node = head;
        while (node != null) {
            node.next = heap.poll();
            node = node.next;
        }
        
        return head.next;
    }
}
